### DISCUSSION TITLE
Change catalog connector REST API

### DISCUSSION THREAD

**Discussion by roytman** [fybrik/fybrik#1431] [Developers]
Current catalog connector [API ](https://fybrik.io/dev/reference/connectors-datacatalog/) is not compliant with REST best [practices](https://medium.com/@mwaysolutions/10-best-practices-for-better-restful-api-cbe81b06f291).
For example:

- it doesn't support versioning
- it starts from the root URI
- it doesn't include the required/modified resource in the URL
- it uses the POST HTTP method for the get asset operation. 
- it doesn't define all possible response codes

I'd suggest using the following API:

-  **GET /v1/catalogs/{catalog_id}/assets/{asset_id}**
OperationType can be passed as a query parameter. 
- **POST /v1/catalogs/{catalog_id}/assets
The CreateAssetRequest parameters will be passed as the request body.
- **DELETE /v1/catalogs/{catalog_id}/assets{asset_id}**

An access token can be passed as an HTTP header in all calls.

This API allows us to support different versions and is better swagger compatible. 

Any comments are very welcome.

**Comment by shlomitk1** [comment]
This issue assumes a dataset is defined by its catalog and asset IDs, which is not the case. 
There is a single string representing a dataset ID. Each connector derives from it the relevant arguments. 
Katalog has a notion of namespace being a catalog ID and Asset name being asset ID. For Egeria we had ServerID and AssetID. For others it could be a JSON string with keys unknown to the manager. Unless we change the datasetID to be "catalogId/assetID" for all catalogs, I don't see how the suggested APIs could be applied from within the manager.

**Comment by roytman** [comment]
The mentioned examples (plus WKC) have some combinations of catalogID and assetID.  
I'm not sure what do you mean by "For others, it could be a JSON string with keys unknown to the manager. "At the end of the day, the manager should be able to connect to the catalog.  I assume that other catalogs use REST API, so it should work. Plus, additional tags can be passed as queries parameters.

**Reply by shlomitk1** [reply]
Catalogs may have combinations of catalogID and assetID. However, datasetID used in fybrikApplication does not have them. It is a single string that can be a JSON containing keys x and y. Each connector parses this string differently based on some sort of a convention. Manager does not validate or parse it. Either we replace the existing datasetID with two fields: assetID and catalogID or agree on the same convention for ALL catalogs, e.g. catalogID/assetID. Otherwise, we can't proceed. 
The statement "the manager should be able to connect to the catalog" is not accurate. Manager does not communicate with the catalog directly. For this, we have a catalog-connector. Till now, it was the catalog-connector's task to parse the sent string and retrieve the required fields such as asset and catalog IDs, not the manager's.

**Reply by roytman** [reply]
I see your point
